src/crimson/os/seastore/journal.h

```diff
@@ -0,0 +1,254 @@
+
+using journal_seq_t = uint64_t;
+static constexpr journal_seq_t NO_DELTAS =
+  std::numeric_limits<journal_seq_t>::max();
+
+/**
+ * Segment header
+ *
+ * Every segment contains and encode segment_header_t in the first block.
+ * Our strategy for finding the journal replay point is:
+ * 1) Find the segment with the highest journal_segment_seq
+ * 2) Scan forward from committed_journal_lb to find the most recent
+ *    journal_commit_lb record
+ * 3) Replay starting at the most recent found journal_commit_lb record
+ */
+struct segment_header_t {
+  segment_seq_t journal_segment_seq;
+  segment_id_t physical_segment_id; // debugging
+
+  paddr_t journal_replay_lb;
+
+  DENC(segment_header_t, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.journal_segment_seq, p);
+    denc(v.physical_segment_id, p);
+    denc(v.journal_replay_lb, p);
+    DENC_FINISH(p);
+  }
+};
+
+struct record_header_t {
+  // Fixed portion
+  segment_off_t mdlength;       // block aligned, length of metadata
+  segment_off_t dlength;        // block aligned, length of data
+  journal_seq_t seq;            // current journal seqid
+  checksum_t    full_checksum;  // checksum for full record (TODO)
+  size_t deltas;                // number of deltas
+  size_t extents;               // number of extents
+
+  DENC(record_header_t, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.mdlength, p);
+    denc(v.dlength, p);
+    denc(v.seq, p);
+    denc(v.full_checksum, p);
+    denc(v.deltas, p);
+    denc(v.extents, p);
+    DENC_FINISH(p);
+  }
+};
+
+/**
+ * Callback interface for managing available segments
+ */
+class JournalSegmentProvider {
+public:
+  using get_segment_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using get_segment_ret = get_segment_ertr::future<segment_id_t>;
+  virtual get_segment_ret get_segment() = 0;
+
+  /* TODO: we'll want to use this to propogate information about segment contents */
+  virtual void put_segment(segment_id_t segment) = 0;
+
+  virtual ~JournalSegmentProvider() {}
+};
+
+/**
+ * Manages stream of atomically written records to a SegmentManager.
+ */
+class Journal {
+public:
+  Journal(SegmentManager &segment_manager);
+
+  /**
+   * Sets the JournalSegmentProvider.
+   *
+   * Not provided in constructor to allow the provider to not own
+   * or construct the Journal (TransactionManager).
+   *
+   * Note, Journal does not own this ptr, user must ensure that
+   * *provider outlives Journal.
+   */
+  void set_segment_provider(JournalSegmentProvider *provider) {
+    segment_provider = provider;
+  }
+
+  /**
+   * initializes journal for new writes -- must run prior to calls
+   * to submit_record.  Should be called after replay if not a new
+   * Journal.
+   */
+  using init_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error
+    >;
+  init_ertr::future<> open_for_write();
+
+  /**
+   * close journal
+   *
+   * TODO: should probably flush and disallow further writes
```

Comment:
seems a must, as roll_journal_segment() calls 'close()' but does nothing
else to guarantee no more writes, isn't it?

Reply:
It's a todo because the details as far as Journal is concerned depend on exactly how the close/shutdown sequence works for the upstream interfaces.  No point in investing effort in it now as it will be (relatively) easy to retrofit later.